src/engine/SortedUnionImpl.h

```diff
@@ -0,0 +1,241 @@
+  bool aggregatedTableReturned_ = false;
+  // Function forwarded from `Union`
+  Func applyPermutation_;
+
+  SortedUnionImpl(std::shared_ptr<const Result> result1,
+                  std::shared_ptr<const Result> result2, Range1 range1,
+                  Range2 range2, bool requestLaziness,
+                  const std::vector<std::array<size_t, 2>>& columnOrigins,
+                  const ad_utility::AllocatorWithLimit<Id>& allocator,
+                  std::vector<ColumnIndex> leftPermutation,
+                  std::vector<ColumnIndex> rightPermutation,
+                  std::span<const ColumnIndex, SPAN_SIZE> comparatorView,
+                  Func applyPermutation)
+      : result1_{std::move(result1)},
+        result2_{std::move(result2)},
+        range1_{std::move(range1)},
+        range2_{std::move(range2)},
+        resultTable_{columnOrigins.size(), allocator},
+        allocator_{allocator},
+        requestLaziness_{requestLaziness},
+        columnOrigins_{columnOrigins},
+        leftPermutation_{std::move(leftPermutation)},
+        rightPermutation_{std::move(rightPermutation)},
+        applyPermutation_{std::move(applyPermutation)} {
+    if (requestLaziness) {
+      resultTable_.reserve(Union::chunkSize);
+    }
+    targetOrder_.reserve(comparatorView.size());
+    for (auto& col : comparatorView) {
+      targetOrder_.push_back(columnOrigins_.at(col));
+    }
+  }
+
+  // Always inline makes makes a huge difference on large datasets.
+  AD_ALWAYS_INLINE bool isSmaller(const auto& row1, const auto& row2) const {
+    using StaticRange = std::span<const std::array<size_t, 2>, SPAN_SIZE>;
+    for (auto [index1, index2] : StaticRange{targetOrder_}) {
+      if (index1 == Union::NO_COLUMN) {
+        return true;
+      }
+      if (index2 == Union::NO_COLUMN) {
+        return false;
+      }
+      if (row1[index1] != row2[index2]) {
+        return row1[index1] < row2[index2];
+      }
+    }
+    return false;
+  }
+
+  // Fetch the next element from the range, make a copy if it's a `Wrapper`.
+  // Otherwise move it. If the range is exhausted, return `std::nullopt`.
+  std::optional<Result::IdTableVocabPair> passNext(
+      const std::vector<ColumnIndex>& permutation, auto& it, auto end,
+      size_t& index) {
+    if (it != end) {
+      Result::IdTableVocabPair pair = moveOrCopy(*it);
+      pair.idTable_ = applyPermutation_(std::move(pair.idTable_), permutation);
+
+      index = 0;
+      ++it;
+      return pair;
+    }
+    return std::nullopt;
+  }
+
+  // Append the remainder of the last partially consumed table to the current
+  // result table and merge the local vocabs.
+  void appendCurrent(const std::vector<ColumnIndex>& permutation, auto& it,
+                     size_t& index) {
+    resultTable_.insertAtEnd(it->idTable_, index, std::nullopt, permutation,
+                             Id::makeUndefined());
+    localVocab_.mergeWith(std::span{&it->localVocab_, 1});
```

Comment:
Suggestion for a small cleanup PR
: overload `LocalVocab::mergeWith(aSingleLocalVocab)` 
to reduce the visual overhead of this `span of single value` pattern which we have all over the place.